spv/verifypayment.go

```diff
@@ -7,30 +7,91 @@ import (
 )
 
-// VerifyPayment verifies whether or not the txs supplied via the supplied spv.Envelope are valid
-func (v *verifier) VerifyPayment(ctx context.Context, initialPayment *Envelope) (bool, error) {
+// VerifyPayment verifies whether or not the txs supplied via the supplied spv.Envelope are valid.
+// This method also accepts functional options which will override the options set in the verifier upon creation.
+// Any options passed here persist only for that call and do not override the main options. This gives flexibility
+// in that you can setup a verifier with sensible defaults and override them conditionally by providing
+// options here to enable or disable checks as required.
+//
+//  // with functional options to override verifier defaults.
+//  tx, err := VerifyPayment(ctx, env, spv.VerifyProofs(), spv.VerifyFees(fees))
+//
+//  // without functional options to use verifier defaults.
+//  tx, err := VerifyPayment(ctx, env)
+//
+func (v *verifier) VerifyPayment(ctx context.Context, initialPayment *Envelope, opts ...VerifyOpt) (*bt.Tx, error) {
 	if initialPayment == nil {
-		return false, ErrNilInitialPayment
+		return nil, ErrNilInitialPayment
+	}
+	vOpt := v.opts.clone()
+	for _, opt := range opts {
+		opt(vOpt)
+	}
+	if vOpt.proofs && v.bhc == nil {
+		return nil, errors.New("at least one blockchain header implementation should be set in the verifier if validating proofs")
+	}
+	// parse initial tx, fail fast if it isn't a valid tx.
+	tx, err := bt.NewTxFromString(initialPayment.RawTx)
+	if err != nil {
+		return nil, err
 	}
 
-	// The tip tx is the transaction we're trying to verify, and it should not have a supplied
-	// Merkle Proof.
-	if initialPayment.IsAnchored() {
-		return false, ErrTipTxConfirmed
+	// verify tx fees
+	if vOpt.fees {
+		if err := v.verifyFees(initialPayment, tx, vOpt); err != nil {
+			return nil, err
+		}
+	}
+	// if we are validating proofs or scripts carry out the validation.
+	if vOpt.requiresEnvelope() {
+		// The tip tx is the transaction we're trying to verify, and it should not have a supplied
+		// Merkle Proof.
+		if initialPayment.IsAnchored() {
+			return nil, ErrTipTxConfirmed
+		}
+		if err := v.verifyTxs(ctx, initialPayment, vOpt); err != nil {
+			return nil, err
+		}
 	}
+	return tx, nil
+}
 
-	valid, err := v.verifyTxs(ctx, initialPayment)
+// verifyFees takes the initial payment and iterates the immediate parents in order to gather
+// the satoshis used for each input of the initialPayment tx.
+//
+// If there are no parents the method will fail, also, if there are no fees the method will fail.
+func (v *verifier) verifyFees(initialPayment *Envelope, tx *bt.Tx, opts *verifyOptions) error {
+	if initialPayment.Parents == nil || len(initialPayment.Parents) == 0 {
+		return ErrCannotCalculateFeePaid
+	}
+	if opts.feeQuote == nil {
+		return ErrNoFeeQuoteSupplied
+	}
+	for _, input := range tx.Inputs {
+		pTx, err := bt.NewTxFromString(initialPayment.Parents[input.PreviousTxIDStr()].RawTx)
```

Comment:
Could this also be split out maybe, so we have can have a "NotAllInputsSupplied" error case, similar to here https://github.com/libsv/go-bc/blob/master/spv/verifypayment.go#L102